core/src/main/java/com/digitalpebble/stormcrawler/protocol/file/FileResponse.java

```diff
@@ -0,0 +1,193 @@
+import java.util.Date;
+import java.util.Locale;
+
+import org.apache.http.HttpStatus;
+import org.slf4j.LoggerFactory;
+
+import com.digitalpebble.stormcrawler.Metadata;
+import com.digitalpebble.stormcrawler.protocol.HttpHeaders;
+import com.digitalpebble.stormcrawler.protocol.ProtocolResponse;
+
+public class FileResponse {
+
+    /**
+     * Adopted from Apache Nutch File Protocol implementation
+     * 
+     * @see https://github.com/apache/nutch/blob/master/src/plugin/protocol-file
+     *      /src/java/org/apache/nutch/protocol/file/FileResponse.java
+     * @see https://github.com/apache/nutch/blob/master/src/plugin/protocol-file
+     *      /src/java/org/apache/nutch/protocol/file/File.java
+     * 
+     **/
+
+    private byte[] content;
+    private int statusCode;
+    private Metadata metadata;
+    private final FileProtocol fileProtocol;
+    static final SimpleDateFormat dateFormat = new SimpleDateFormat(
+            "EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US);
+    static final org.slf4j.Logger LOG = LoggerFactory.getLogger(MethodHandles
+            .lookup().lookupClass());
+
+    public FileResponse(String u, Metadata md, FileProtocol fp)
+            throws MalformedURLException, IOException {
+
+        fileProtocol = fp;
+        metadata = md;
+
+        URL url = new URL(u);
+
+        if (!url.getPath().equals(url.getFile())) {
+            LOG.warn("url.getPath() != url.getFile(): {}.", url);
+        }
+
+        String path = "".equals(url.getPath()) ? "/" : url.getPath();
+
+        File file = new File(
+                URLDecoder.decode(path, fileProtocol.getEncoding()));
+
+        if (!file.exists()) {
+            statusCode = HttpStatus.SC_NOT_FOUND;
+            return;
+        }
+
+        if (!file.canRead()) {
+            statusCode = HttpStatus.SC_UNAUTHORIZED;
+            return;
+        }
+
+        if (!file.equals(file.getCanonicalFile())) {
+            metadata.setValue(HttpHeaders.LOCATION, file.getCanonicalFile()
+                    .toURI().toURL().toString());
+            statusCode = HttpStatus.SC_MULTIPLE_CHOICES;
+            return;
+        }
+
+        if (file.isDirectory()) {
+            getDirAsHttpResponse(file);
+        } else if (file.isFile()) {
+            getFileAsHttpResponse(file);
+        } else {
+            statusCode = HttpStatus.SC_INTERNAL_SERVER_ERROR;
+            return;
+        }
+
+        if (content == null) {
+            content = new byte[0];
+        }
+    }
+
+    public ProtocolResponse toProtocolResponse() throws IOException {
+        return new ProtocolResponse(content, statusCode, metadata);
+    }
+
+    private void getFileAsHttpResponse(File file) throws IOException {
+        long size = file.length();
+
+        if (size > Integer.MAX_VALUE) {
+            statusCode = HttpStatus.SC_BAD_REQUEST;
+            return;
+        }
+
+        int len = (int) size;
+
+        if (fileProtocol.getMaxContentLength() >= 0

```

Comment:
not sure we need to enforce the checks on the max length. This is useful for http as we don't want to fetch streams of audio or video indefinitely but for files this should not be a problem.

We could instead use the utils from [commons-io]( 
https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/IOUtils.html#toByteArray(java.io.InputStream)) to simplify our code. I thinkthere is a variant of the method which takes the max length.